### PR TITLE
nixos: allow customizing the kernel RANDSTRUCT seed

### PR DESCRIPTION
**Pull request by delroth**
###### Motivation for this change

See #53592. RANDSTRUCT is designed to make kernel exploitation harder through diversification of the builds. Allowing users to set their own RANDSTRUCT seed differentiates their kernel build from the rest of the NixOS hardened users. In a multi-host deployment this can also be used to deploy differently seeded builds on each machine.

I was torn on whether to put this in boot/kernel.nix or create a new module in security/ for the option -- in the end I decided on boot/kernel.nix because it makes the implementation much simpler, but I'm definitely open to the "new module" option too if someone has a strong opinion about this.

Still a NixOS noob, not sure if this is the right way to do things. Let me know!

@joachifm @NeQuissimus fyi

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by joachifm** [comment]
Nit:  Exposing such a specialized feature in the general interface may seem a little ad-hoc.
Are we certain that the seed is only ever used for randstruct? If not, I could see adding it to the general interface, otherwise I'd prefer having this under the hardened module namespace.

File: nixos/modules/system/boot/kernel.nix [hunk 1]
```diff
@@ -67,6 +68,19 @@ in
       description = "A list of additional patches to apply to the kernel.";
     };
 
+    boot.kernel.randstructSeed = mkOption {
+      type = types.str;
```
Comment (joachifm) [inline nixos/modules/system/boot/kernel.nix]: It'd be neat to be able to read the seed from a file instead (e.g., if I want to share my config but not the seed)

**Comment by joachifm** [comment]
If you don't want to create a new module, there's `security/misc`

File: pkgs/os-specific/linux/kernel/manual-config.nix [hunk 2]
```diff
@@ -111,7 +114,7 @@ let
         if [ -f scripts/gcc-plugins/gen-random-seed.sh ]; then
           substituteInPlace scripts/gcc-plugins/gen-random-seed.sh \
             --replace NIXOS_RANDSTRUCT_SEED \
-            $(echo ${src} ${configfile} | sha256sum | cut -d ' ' -f 1 | tr -d '\n')
+            $(echo ${randstructSeed} ${src} ${configfile} | sha256sum | cut -d ' ' -f 1 | tr -d '\n')
```
Comment (danbst) [inline pkgs/os-specific/linux/kernel/manual-config.nix]: you may remove space in `${randstructSeed} ${src}` and this won't trigger kernel (and pkgs) rebuild. Maybe like
```
${lib.concatMapStringsSep " " builtins.toString [randstructSeed src configfile]}
``` 
or maybe just `${configfile}` -> `${randstructSeed + configfile}`

**Comment by joachifm** [comment]
@GrahamcOfBorg build linux_latest_hardened linux linux_latest

**Comment by joachifm** [comment]
@GrahamcOfBorg test hardened

**Pull request merged**